src/algorithms/validation/check.cpp

```diff
@@ -7,8 +7,10 @@ All rights reserved (see LICENSE).
 
 */
 
+#include <algorithm>
```

Comment:
About those "missing" includes: my understanding is that we're kind of fine when another internal include does have the required ones. For example, most files end up including `typedefs.h` (directly or not) so they actually have `optional`, `string`, `vector` at hand. This assumption is probably a bit fuzzy though since we've had reports of compilation problems with such "missing" includes in the past.

Did you spot those based on compilation errors with some toolchain? Or is it just a common good practice to explicitly include everything? If so did you use a specific tool to spot the missing ones?

Reply:
Well, one could argue that it is ok as long as the code compiles and it does :-) On the other hand if you remove an include from some header file (e.g. `typedefs.h`) due to a refactor, this could case compilation errors in some source files that depend on that include but do not include themselves. This can become tedious to fix.

I think it is a good practice to explicitly include everything, and there are tools to help with this, like [include-what-you-use](https://include-what-you-use.org/) (which by the way has a [rationale](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md)).

Anyway, the missing includes I added are result of eyeballing the codebase and are an ad-hoc effort. I did not use any tool.

Reply:
Thanks for the pointers, makes total sense. I'd be in favor of running a more general cleanup campaign, which I ticketed in #797.